src/Adapter/Product/Shop/CommandHandler/DeleteProductFromShopsHandler.php

```diff
@@ -76,4 +89,27 @@ public function handle(DeleteProductFromShopsCommand $command): void
 
         $this->productRepository->deleteFromShops($command->getProductId(), $shopIds);
     }
+
+    /**
+     * @param ProductId $productId
+     * @param ShopId[] $shopIds
+     */
+    private function removeImages(ProductId $productId, array $shopIds): void
+    {
+        // first collect all the images from shops and make sure they are not duplicating
+        $imageIds = [];
+        foreach ($shopIds as $shopId) {
+            $imageIds = array_merge(
+                $imageIds,
+                array_map(static function (ImageId $imageId): int {
+                    return $imageId->getValue();
+                }, $this->productImageMultiShopRepository->getImagesIds($productId, ShopConstraint::shop($shopId->getValue())))
+            );
+        }
+
+        // then remove each of those images from all the specified shops
+        foreach ($imageIds as $imageId) {
+            $this->productImageMultiShopRepository->deleteFromShops(new ImageId($imageId), $shopIds);
+        }
+    }
```

Comment:
I think you should handle the removal in the same loop shop by shop Because in `deleteFromShops` you check the association between the ImageId and the ShopId But here you merge all the imageIds from all the shops, so you could end up with non existing association and some error will trigger

Reply:
```suggestion
        foreach ($shopIds as $shopId) {
            $imageIds = $this->productImageMultiShopRepository->getImagesIds($productId, ShopConstraint::shop($shopId->getValue()));
            foreach ($imageIds as $imageId) {
                $this->productImageMultiShopRepository->deleteFromShops(new ImageId($imageId), [$shopId]);
            }
        }
    }
```